pilot/pkg/networking/grpcgen/cds.go

```diff
@@ -0,0 +1,216 @@
+
+import (
+	"fmt"
+	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	"github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	tlsv3 "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/core/v1alpha3"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/config/host"
+)
+
+// BuildClusters handles a gRPC CDS request, used with the 'ApiListener' style of requests.
+// The main difference is that the request includes Resources to filter.
+func (g *GrpcConfigGenerator) BuildClusters(node *model.Proxy, push *model.PushContext, names []string) model.Resources {
+	var clusters = make([]*clusterv3.Cluster, 0, len(names))
+	for _, defaultClusterName := range names {
+		builder, err := newClusterBuilder(node, push, defaultClusterName)
+		if err != nil {
+			log.Warn(err)
+			continue
+		}
+		clusters = append(clusters, builder.build()...)
+	}
+
+	resp := make(model.Resources, 0, len(clusters))
+	for _, c := range clusters {
+		resp = append(resp, &envoy_service_discovery_v3.Resource{
+			Name:     c.Name,
+			Resource: util.MessageToAny(c),
+		})
+	}
+	return resp
+}
+
+// clusterBuilder is responsible for building a single default and subset clusters for a service
+type clusterBuilder struct {
+	// conveinence
+	push *model.PushContext
+	node *model.Proxy
+
+	// guaranteed to be set in init
+	name           string
+	defaultCluster *clusterv3.Cluster
+	hostname       host.Name
+	portNum        int
+
+	// may not be set
+	svc  *model.Service
+	port *model.Port
+}
+
+func newClusterBuilder(node *model.Proxy, push *model.PushContext, defaultClusterName string) (*clusterBuilder, error) {
+	_, _, hostname, portNum := model.ParseSubsetKey(defaultClusterName)
+	if hostname == "" || portNum == 0 {
+		return nil, fmt.Errorf("failed parsing subset key: %s", defaultClusterName)
+	}
+
+	defaultCluster := edsCluster(defaultClusterName)
+	if defaultCluster == nil {
+		// should never happen
+		return nil, fmt.Errorf("failed building default cluster %s", defaultClusterName)
+	}
+
+	// try to resolve the service and port
+	var port *model.Port
+	svc := push.ServiceForHostname(node, hostname)
+	if svc == nil {
+		log.Warnf("cds gen for %s: did not find service for cluster %s", node.ID, defaultClusterName)
+	} else {
+		var ok bool
+		port, ok = svc.Ports.GetByPort(portNum)
+		if !ok {
+			log.Warnf("cds gen for %s: did not find port %d in service for cluster %s", node.ID, portNum, defaultClusterName)
+		}
+	}
+
+	return &clusterBuilder{
+		node:           node,
+		push:           push,
+		hostname:       hostname,
+		portNum:        portNum,
+		defaultCluster: defaultCluster,
+		svc:            svc,
+		port:           port,
+	}, nil
+}
+
+func (b *clusterBuilder) build() []*clusterv3.Cluster {
```

Comment:
Most of the logic is similar, I think we can just share the code if we expose:
* BuildSubsetCluster
* BuildDefaultCluster
* BuildClusterOpts and members
* Add something to allow us to override how tlscontext is built

Reply:
i think we should do this before merge